eventuate-messaging-kafka-basic-consumer/src/test/java/io/eventuate/messaging/kafka/basic/consumer/EventuateKafkaConsumerTest.java

```diff
@@ -0,0 +1,152 @@
+package io.eventuate.messaging.kafka.basic.consumer;
+
+import io.eventuate.messaging.kafka.producer.EventuateKafkaProducer;
+import io.eventuate.messaging.kafka.producer.EventuateKafkaProducerConfigurationProperties;
+import io.eventuate.util.test.async.Eventually;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.boot.context.properties.EnableConfigurationProperties;
+import org.springframework.boot.test.context.SpringBootTest;
+import org.springframework.test.context.junit4.SpringRunner;
+
+import java.util.Collections;
+import java.util.UUID;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.TimeUnit;
+
+@RunWith(SpringRunner.class)
+@SpringBootTest(classes = EventuateKafkaConsumerTest.Config.class)
+public class EventuateKafkaConsumerTest {
+
+  @EnableConfigurationProperties({EventuateKafkaConsumerConfigurationProperties.class,
+          EventuateKafkaProducerConfigurationProperties.class})
+  public static class Config {
+  }
+
+  @Value("${eventuatelocal.kafka.bootstrap.servers}")
+  private String bootstrapServers;
+
+  @Autowired
+  private EventuateKafkaConsumerConfigurationProperties eventuateKafkaConsumerConfigurationProperties;
+
+  @Autowired
+  private EventuateKafkaProducerConfigurationProperties eventuateKafkaProducerConfigurationProperties;
+
+  private EventuateKafkaProducer eventuateKafkaProducer;
+
+  private volatile boolean recordHandled = false;
+
+  private String subscriberId = UUID.randomUUID().toString();
+  private String topic = UUID.randomUUID().toString();
+
+  private LinkedBlockingQueue<String> queue = new LinkedBlockingQueue<>();
+
+  @Before
+  public void init() {
+    eventuateKafkaProducer = new EventuateKafkaProducer(bootstrapServers, eventuateKafkaProducerConfigurationProperties);
+
+    subscriberId = UUID.randomUUID().toString();
+    topic = UUID.randomUUID().toString();
+    recordHandled = false;
```

Comment:
Can't this be implemented using a Mockito mock?

Reply:
Could you please clarify what exactly?

Reply:
`recordHandled` is used to indicate that the `EventuateKafkaConsumerMessageHandler`.
That's exactly what 

```
handler = mock(EventuateKafkaConsumerMessageHandler.class), 

verify(mock)....
```

does.

Reply:
Chris,

all tests has that check, but in some tests handler does also something else.
But there is a common method for all cases which checks that handler is invoked.
Mocking handlers when important only fact of invocation will require separate method and more code. I don't think that mocking can simplify these tests. I think it is only a complication.

Reply:
```
private void assertRecordHandled() {
    Eventually.eventually(() -> verify(handler);
  }
```
```
when(handler.apply(any()), any()).thenThrow()
```

Reply:
ok, but we still have this case:

```
    createConsumer((record, callback) -> {
      callback.accept(null, new RuntimeException("Something happend"));
      recordHandled = true;
      return null;
    });
```

it does not throw exception, it sends it callback

Reply:
Please read the manual:

```
    when(handler.apply(any()), any()).then(invocation -> {
      invocation.getArgument(...)
      return null;
    });
```

Reply:
oh, ok, thank you